src/classifier/SparseLabeledParameterVector.h

```diff
@@ -98,6 +103,34 @@ class SparseLabelWeights : public LabelWeights {
     }
     label_weights_.push_back(std::pair<int, double>(label, weight));
   }
+  // Sets new weight, normalize it and returns previous value of label_weights_[k].second, with k such that label == label_weights_[k].first
+  double SetWeightAndNormalize(int label, double value, double scaling_factor) {
```

Comment:
Note: these optimizations in Set/AddWeightAndNormalize are fine, but they won't apply in test time. At test time, the weight vector is fixed so these functions are never called. This is only speeding up training.